--- gawk-3.1.2/po/Makefile.in.in.orig	Mon May 26 12:51:37 2003
+++ gawk-3.1.2/po/Makefile.in.in	Mon May 26 12:51:40 2003
@@ -27,7 +27,7 @@
 INSTALL = @INSTALL@
 INSTALL_DATA = @INSTALL_DATA@
 MKINSTALLDIRS = @MKINSTALLDIRS@
-mkinstalldirs = $(SHELL) `case "$(MKINSTALLDIRS)" in /*) echo "$(MKINSTALLDIRS)" ;; *) echo "$(top_builddir)/$(MKINSTALLDIRS)" ;; esac`
+mkinstalldirs = $(SHELL) `case "$(MKINSTALLDIRS)" in /*) echo "$(MKINSTALLDIRS)" ;; *) echo "$(MKINSTALLDIRS)" ;; esac`
 
 GMSGFMT = @GMSGFMT@
 MSGFMT = @MSGFMT@
