diff -urN aa/gawk-3.1.3/acinclude.m4 gawk-3.1.3/acinclude.m4
--- aa/gawk-3.1.3/acinclude.m4	2002-10-06 21:43:58.000000000 +0200
+++ gawk-3.1.3/acinclude.m4	2004-02-01 14:53:46.007338912 +0100
@@ -1,6 +1,6 @@
-m4_sinclude(arch.m4)dnl
-m4_sinclude(gettext.m4)dnl
-m4_sinclude(lcmessage.m4)dnl
-m4_sinclude(progtest.m4)dnl
-m4_sinclude(socket.m4)dnl
-m4_sinclude(strtod.m4)dnl
+m4_sinclude(m4/arch.m4)dnl
+m4_sinclude(m4/gettext.m4)dnl
+m4_sinclude(m4/lcmessage.m4)dnl
+m4_sinclude(m4/progtest.m4)dnl
+m4_sinclude(m4/socket.m4)dnl
+m4_sinclude(m4/strtod.m4)dnl
