diff -Nru gawk-3.1.2.orig/Makefile.am gawk-3.1.2/Makefile.am
--- gawk-3.1.2.orig/Makefile.am	Thu Mar  6 10:31:38 2003
+++ gawk-3.1.2/Makefile.am	Fri Apr  4 13:00:43 2003
@@ -112,7 +112,7 @@
 pgawk_SOURCES = $(base_sources) eval_p.c profile_p.c
 
 # Get extra libs as needed
-LDADD = @LIBINTL@ @SOCKET_LIBS@
+LDADD = @LIBINTL@
 
 # Set autoconf Makefile.in variables correctly
 datadir = @datadir@
diff -Nru gawk-3.1.2.orig/configure.in gawk-3.1.2/configure.in
--- gawk-3.1.2.orig/configure.in	Sun Mar 16 11:23:32 2003
+++ gawk-3.1.2/configure.in	Fri Apr  4 13:01:47 2003
@@ -23,6 +23,12 @@
 
 dnl Process this file with autoconf to produce a configure script.
 
+m4_pattern_allow(GAWK_AC_LINUX_ALPHA)
+m4_pattern_allow(GAWK_AC_TYPE_SSIZE_T)
+m4_pattern_allow(GAWK_AC_FUNC_STRTOD_C89)
+m4_pattern_allow(GAWK_AC_LIB_SOCKETS)
+m4_pattern_allow(GAWK_AC_AIX_TWEAK)
+
 dnl *** IMPORTANT ***: Fix version in both AC_INIT and AM_INIT_AUTOMAKE
 AC_INIT([GNU Awk], 3.1.2, bug-gawk@gnu.org, gawk)
 
