--- gawk-5.0.0/test/Makefile.am.orig	2019-04-07 20:53:42.000000000 +0200
+++ gawk-5.0.0/test/Makefile.am	2019-04-12 21:10:26.219201910 +0200
@@ -934,10 +934,6 @@
 	prt1eval.ok \
 	prtoeval.awk \
 	prtoeval.ok \
-	pty1.awk \
-	pty1.ok \
-	pty2.awk \
-	pty2.ok \
 	rand-mpfr.ok \
 	rand.awk \
 	rand.ok \
@@ -1498,8 +1498,8 @@
 	octdec patsplit posix printfbad1 printfbad2 printfbad3 printfbad4 \
 	printhuge procinfs profile0 profile1 profile2 profile3 profile4 \
 	profile5 profile6 profile7 profile8 profile9 profile10 profile11 \
-	profile12 profile13 profile14 profile15 profile16 profile17 pty1 \
-	pty2 rebuf regexsub reginttrad regnul1 regnul2 regx8bit reint \
+	profile12 profile13 profile14 profile15 profile16 profile17 \
+	rebuf regexsub reginttrad regnul1 regnul2 regx8bit reint \
 	reint2 rsgetline rsglstdin rsstart1 rsstart2 rsstart3 rstest6 \
 	sandbox1 shadow shadowbuiltin sortfor sortfor2 sortu sourcesplit \
 	split_after_fpat splitarg4 strftfld strftime strtonum strtonum1 \
