--- gawk-4.1.4/test/Makefile.am.orig	2016-08-24 21:31:55.000000000 +0200
+++ gawk-4.1.4/test/Makefile.am	2016-08-30 22:03:59.248680327 +0200
@@ -780,8 +780,6 @@
 	prt1eval.ok \
 	prtoeval.awk \
 	prtoeval.ok \
-	pty1.awk \
-	pty1.ok \
 	rand-mpfr.ok \
 	rand.awk \
 	rand.ok \
@@ -1126,7 +1124,7 @@
 	nastyparm negtime next nondec nondec2 \
 	patsplit posix printfbad1 printfbad2 printfbad3 printfbad4 printhuge procinfs \
 	profile0 profile1 profile2 profile3 profile4 profile5 profile6 \
-	profile7 profile8 pty1 \
+	profile7 profile8 \
 	rebuf regnul1 regnul2 regx8bit reginttrad reint reint2 rsgetline rsglstdin rsstart1 \
 	rsstart2 rsstart3 rstest6 shadow sortfor sortfor2 sortu split_after_fpat \
 	splitarg4 strftime \
