--- gawk-4.1.2/test/Makefile.am.orig	2015-04-28 15:49:46.000000000 +0200
+++ gawk-4.1.2/test/Makefile.am	2015-05-10 07:28:01.671200070 +0200
@@ -743,8 +743,6 @@
 	prt1eval.ok \
 	prtoeval.awk \
 	prtoeval.ok \
-	pty1.awk \
-	pty1.ok \
 	rand-mpfr.ok \
 	rand.awk \
 	rand.ok \
@@ -1060,7 +1058,7 @@
 	manyfiles match1 match2 match3 mbstr1 \
 	nastyparm negtime next nondec nondec2 \
 	patsplit posix printfbad1 printfbad2 printfbad3 printfbad4 printhuge procinfs \
-	profile0 profile1 profile2 profile3 profile4 profile5 profile6 profile7 pty1 \
+	profile0 profile1 profile2 profile3 profile4 profile5 profile6 profile7 \
 	rebuf regnul1 regnul2 regx8bit reginttrad reint reint2 rsgetline rsglstdin rsstart1 \
 	rsstart2 rsstart3 rstest6 shadow sortfor sortu split_after_fpat \
 	splitarg4 strftime \
